infra/src/main/kotlin/fi/fta/geoviite/api/frameconverter/v1/FrameConverterControllerV1.kt

```diff
@@ -1,13 +1,84 @@
+import fi.fta.geoviite.api.frameconverter.geojson.GeoJsonFeatureCollection
+import fi.fta.geoviite.infra.common.MainLayoutContext
+import fi.fta.geoviite.infra.error.IntegrationApiException
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.http.converter.json.Jackson2ObjectMapperBuilder
+import org.springframework.web.bind.annotation.ModelAttribute
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RequestMethod
+import org.springframework.web.bind.annotation.RequestParam
 
-@GeoviiteIntegrationApiController("/frame-converter/v1")
-class FrameConverterControllerV1 {
+@Configuration
+class FrameConverterApiObjectMapperV1 {
 
-    @GetMapping("/some-conversion")
-    fun someConversion(): String {
-        return "some response"
+    @Bean
+    fun objectMapper(builder: Jackson2ObjectMapperBuilder): ObjectMapper {
+        val module = SimpleModule()
+        module.addDeserializer(FrameConverterRequestV1::class.java, FrameConverterRequestDeserializerV1())
+        return builder.build<ObjectMapper>().registerModule(module)
+    }
+}
+
+@GeoviiteIntegrationApiController(
+    [
+        "/rata-vkm",
+        "/rata-vkm/dev",
+        "/rata-vkm/v1",
+        "/rata-vkm/dev/v1",
+    ],
+)
+class FrameConverterControllerV1 @Autowired constructor(
+    private val frameConverterServiceV1: FrameConverterServiceV1,
+) {
+
+    @RequestMapping(method = [RequestMethod.GET, RequestMethod.POST])
+    fun multiInputTransform(
+        @RequestParam(required = false, name = "json") requests: List<FrameConverterRequestV1>?,
+    ): GeoJsonFeatureCollection {
+        val parsedRequests = requests ?: emptyList()
+
+        return GeoJsonFeatureCollection(
+            features = parsedRequests.flatMap(::processRequest),
+        )
+    }
+
+    @RequestMapping(method = [RequestMethod.GET, RequestMethod.POST], params = ["!json"])
+    fun singleInputTransform(
+        @ModelAttribute request: CoordinateToTrackMeterRequestV1,
+    ): GeoJsonFeatureCollection {
+        return GeoJsonFeatureCollection(
+            features = processRequest(request),
+        )
+    }
+
+    private fun processRequest(request: FrameConverterRequestV1): List<GeoJsonFeature> {
+        return when (request) {
+            is CoordinateToTrackMeterRequestV1 -> {
+                val (validatedRequest, errorResponse)
+                    = frameConverterServiceV1.validateCoordinateToTrackMeterRequest(request)
+
+                if (validatedRequest == null) {
+                    errorResponse
+                } else {
+                    frameConverterServiceV1.coordinateToTrackAddress(
+                        layoutContext = MainLayoutContext.official,
+                        request = validatedRequest,
+                    )
+                }
+            }
```

Comment:
Nämä voisi mielestäni kääntää toisin päin (happy case ensin.) Helpottaisi lukemista kun ensin olisi iloinen tapaus ja virhetilanteet tulisivat vasta sen jälkeen. Nykyisellään iloinen tapaus on vähän sandwichattuna virhekäsittelyjen väliin.

Reply:
Vähän kumsi-kumsa. Early-returneja suosivana errori varmaan päätynyt ylemmäksi, jolloin lopusta on nähtävissä se happycasen päättyminen (standardoidusti aina jos samaa rakennetta aina käyttäisi). Mutta ei tosiaan juurikaan preferenssejä suuntaan tai toiseen, voidaan kyllä vaihtaa.